pip/pip-322.md

```diff
@@ -0,0 +1,143 @@
+# PIP-322: Pulsar Rate Limiting Refactoring
+
+# Motivation
+
+The current rate limiting implementation in Apache Pulsar has several
+known issues that impact performance and accuracy when operating Pulsar
+clusters under load (detailed in [Problems to
+Address](#problems-to-address)). This proposal outlines a refactor to
+consolidate multiple existing options into a single improved solution.
+
+The refactor aims to resolve numerous user complaints regarding default
+Pulsar rate limiting being unusable. In addition, inconsistencies and
+thread contention with existing rate limiters cause unpredictable
+throttling and added latency.
+
+Overhauling the built-in implementation will improve multi-tenancy,
+allow Pulsar to scale to demanding workloads, and address longstanding
+issues.
+
+Rate limiters act as a conduit to more extensive capacity management and
+Quality of Service (QoS) controls in Pulsar. They are integral to
+Pulsar's core multi-tenancy features. This refactoring will pave the way
+for future enhancements.
+
+# Background Reading
```

Comment:
In many previous PIPs submitted by Pulsar community, this technique providing external links the reader must deep dive and read first, proved to make PIP review ultra hard (I can attest to it personally).
The difficulties that I saw with this:
* External links are not guaranteed to stay there. A user in 2-3 years would like to understand this PIP better, but it relies on data which is there anymore
* You can't comment on on external site data
* It takes a very long time to ready something which the author only needed maybe 50% of it. A more efficient approach was to digest it for the reader in exactly the context required.

So based on that, a template was introduced, in which we said that external links should not be used.

Specifically in this case, the blog you wrote is brilliant, and the meeting notes are superb. 
I recommend wholeheartedly, to simply take the relevant information from it, and place it in the appropriate place in the PIP, whether its:
* How rate limiters work today - what is this prcecise and non-precise rate limiter.
* What are the problems due to current limitation and why they are created
* High level - What is being removed, and what is being added and how it all works together (as you explained so good in the blog)
* Detailed design - 
   * How the async token implementation works in high level.
   * How the Throttler works, also across connections.

Reply:
I acknowledge that PIPs shouldn't depend on external content. However, I think that in this case, it serves the purpose to provide more context for those who care about it. I don't like link baiting myself either. The gist of external content should be already in the PIP document and I'll try to improve the content to capture it.

Thanks for providing the specific questions. I'll cover those in further updates.

Reply:
Thanks @asafm , I hopefully have these points covered in the most recent version of the PIP document.